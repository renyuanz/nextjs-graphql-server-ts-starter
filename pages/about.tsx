import Link from 'next/link'

import { initializeApollo } from '../lib/apollo'
import {
  useGetViewerQuery,
  GetViewerDocument,
} from '../queries/generated/graphql'

export default function About(): JSX.Element {
  const { data, loading, error } = useGetViewerQuery()

  if (loading) return <div>loading...</div>
  if (error) return <div>{error.message}</div>
  return (
    <div className="container">
      <div>
        You&apos;re signed in as {data?.viewer.name} and you&apos;re{' '}
        {data?.viewer.status} go to the{' '}
        <Link href="/">
          <a>Home</a>
        </Link>{' '}
        page.
      </div>
    </div>
  )
}

export async function getStaticProps() {
  const apolloClient = initializeApollo()

  await apolloClient.query({
    query: GetViewerDocument,
  })

  return {
    props: {
      initialApolloState: apolloClient.cache.extract(),
    },
  }
}
