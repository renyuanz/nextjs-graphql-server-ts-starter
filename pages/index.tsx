import Head from 'next/head'
import Link from 'next/link'

import {
  useGetViewerQuery,
  GetViewerDocument,
} from '../queries/generated/graphql'
import { initializeApollo } from '../lib/apollo'

export const Home = (): JSX.Element => {
  const { data, loading, error } = useGetViewerQuery()

  const handleClick = () => window.alert('With typescript and Jest')

  if (loading) return <div>loading...</div>
  if (error) return <div>{error.message}</div>
  return (
    <div className="container">
      <Head>
        <title>Create Next App</title>
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <div>
        You&apos;re signed in as {data?.viewer.name} and you&apos;re{' '}
        {data?.viewer.status} go to the{' '}
        <Link href="/about">
          <a>about</a>
        </Link>{' '}
        page.
        <button onClick={handleClick}>Test Button</button>
      </div>
    </div>
  )
}

export async function getStaticProps() {
  const apolloClient = initializeApollo()

  await apolloClient.query({
    query: GetViewerDocument,
  })

  return {
    props: {
      initialApolloState: apolloClient.cache.extract(),
    },
  }
}

export default Home
