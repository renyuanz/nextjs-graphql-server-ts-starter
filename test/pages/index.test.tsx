import React from 'react'
import { MockedProvider } from '@apollo/client/testing'
import { readFileSync } from 'fs'
import { resolve } from 'path'
import gql from 'graphql-tag'
import { render, fireEvent, wait } from '../testUtils'
import { Home } from '../../pages/index'
import { act } from 'react-dom/test-utils'

const mocks = [
  {
    request: {
      query: gql(
        readFileSync(
          resolve(__dirname, '../../queries/viewer.graphql'),
          'utf-8'
        )
      ),
    },
    result: {
      data: {
        viewer: {
          id: '1',
          name: 'Douglas',
          status: 'cached',
        },
      },
    },
  },
]

const mocksError = [
  {
    request: {
      query: gql(
        readFileSync(
          resolve(__dirname, '../../queries/viewer.graphql'),
          'utf-8'
        )
      ),
    },
    error: new Error('mocks error'),
  },
]

describe('Home page', () => {
  it('matches loading state snapshot', async () => {
    const { asFragment } = render(
      <MockedProvider mocks={mocks}>
        <Home />
      </MockedProvider>
    )

    // test loading state
    expect(asFragment()).toMatchSnapshot()
  })

  it('matches final state snapshot', async () => {
    const { asFragment } = render(
      <MockedProvider mocks={mocks}>
        <Home />
      </MockedProvider>
    )

    // wait 0ms to jump event loop next tick
    await act(async () => {
      await wait(0)
    })

    // test final state
    expect(asFragment()).toMatchSnapshot()
  })

  it('matches error state snapshot', async () => {
    const { asFragment } = render(
      <MockedProvider mocks={mocksError}>
        <Home />
      </MockedProvider>
    )

    // wait 0ms to jump event loop next tick
    await act(async () => {
      await wait(0)
    })

    // test error state
    expect(asFragment()).toMatchSnapshot()
  })

  it('clicking button triggers alert', async () => {
    const { getByText } = render(
      <MockedProvider mocks={mocks}>
        <Home />
      </MockedProvider>
    )

    await act(async () => {
      await wait(0)
    })

    window.alert = jest.fn()
    fireEvent.click(getByText('Test Button'))
    expect(window.alert).toHaveBeenCalledWith('With typescript and Jest')
  })
})
