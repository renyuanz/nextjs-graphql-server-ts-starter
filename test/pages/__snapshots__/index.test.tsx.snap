// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches error state snapshot 1`] = `
<DocumentFragment>
  <div>
    mocks error
  </div>
</DocumentFragment>
`;

exports[`Home page matches final state snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div>
      You're signed in as Douglas and you're cached go to the 
      <a
        href="/about"
      >
        about
      </a>
       page.
      <button>
        Test Button
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Home page matches loading state snapshot 1`] = `
<DocumentFragment>
  <div>
    loading...
  </div>
</DocumentFragment>
`;
